lib/mongo/server/connection_pool/queue.rb

```diff
@@ -149,10 +166,19 @@ def dequeue
         #
         # @since 2.1.0
         def disconnect!
-          check_count_invariants
+          check_count_invariants(false)
           mutex.synchronize do
-            queue.each{ |connection| connection.disconnect! }
-            @pool_size -= queue.length
+            @pool_size -= queue.size
+            queue.each do |connection|
+              connection.disconnect!
+              publish_cmap_event(
```

Comment:
Event publishing invokes application code which may raise exceptions and thus abort the process. For each connection in the queue, that connection must be removed from the queue, pool size reduced by 1, then the event published. Please audit the rest of the diff as well considering effects of exceptions raised during event publishing.

Reply:
Please fix any other occurrences of incorrect pool size management as well.

Reply:
I believe this is all set now